Update the encoding of durations to avoid fidelity loss

danhhz opened the pull request:
Durations were previously encoded the same way golang does, as an int64 count of
nanoseconds. But it's not possible to use this to represent something like "1
year 3 months" without either context or losing some information.

Postgres stores durations as a count of months, days, and fractions of seconds.
This works because there are always 12 months in a year, but the number of days
in a month varies. The number of seconds in a day also varies because of
daylight savings time. When sorting these durations, Postgres assumes that 1
month = 30 days and 1 day = 24 \* 60 \* 60 \* 1E6 nanoseconds. This means that 29d
sorts before 1m which sorts before 31d.

With inspiration from postgres, this change updates the internal representation
of durations to be a count of months, days, and nanoseconds. The encoding used
in sql tables matches the postgres sorting.

bdarnell committed:
testutils: move build.go into a subpackage

This removes a dependency on go/build from the main binary.

dt committed:
server: check for updates

First part of #5295

spencerkimball committed:
Always write intents on WriteTooOldError to prevent starvation

This PR started out with an assumption that this would be easier
than it turned out to be and a lot of changes had to go into making
this a reality. Here's a summary of them.

- On `WriteTooOldError`, MVCC code now writes at the next logical
  timestamp but still returns the error. The error is used to update
  the batch timestamp or batch txn timestamp. This means we can lay
  down intents even when there are busy writes, which otherwise
  were leading to starvation scenarios.
- To deal with the new policy for write too old errors, changes
  were necessary to increment and conditional put to avoid writing
  incorrect values.
- MVCC increment/cput code was NOT correct before for snapshot iso
  because on these two read-then-mutate write commands, we are
  doing the read and the write at the txn timestamp, not the txn
  original timestamp. The conclusion I've come to is that we
  were incorrectly using upgraded txn timestamp for successive
  writes within an epoch of a txn when we should ALWAYS insist
  only on using the `OrigTimestamp` for an epoch.
- Snapshot isolation MUST restart on write too old errors (same
  for serializable isolation, though the diff between timestamp
  and orig timestamp subsumes this). In order to allow snapshot
  isolation transactions to lay down intents, I've introduced a
  new `Transaction.WriteTooOld` boolean in the proto, which is set
  on any `WriteTooOldError` and then consulting during `EndTxn`
  to prompt a restart if the isolation is `SNAPSHOT`.
- MVCC intents are now abandoned instead of committed if they
  have newer timestamp than the resolving transaction. This is an
  integral component of the replay protection and in conjunction
  with the timestamp cache, makes it impossible to commit abandoned
  intents on adverserial replays.
- Thread the txn ID through to the intent resolving so that a txn
  which pushes an intent in its way can proceed without its txn
  timestamp being incremented (and leading to a transaction retry
  for serializable isolation).
- Modify write timestamp cache to only include DeleteRange ops. This
  should be a boon to performance. If we encounter a new timestamp
  from the cache now, we always move the transaction timestamp
  forward and set the WriteTooOld boolean, to guard against phantom
  delete anomalies.
- Removed the "fiddleWithTimestamps" kludge! This cleans up some
  frightening code. Now we just execute the commands in a batch and
  let the timestamp tick up on overlapping write commands. We're
  careful to update the node clock from the store when the batch
  is complete so that subsequent commands are guaranteed to see
  the work done by the batch.

We were previously laboring under the misapprehension that on
`WriteTooOldErrors` we were retrying transactions, thereby
providing replay protection after txn entries and the sequence
cache were cleared. What we were actually doing was retrying
immediately from the store level (so going back through raft and
thereby not avoiding the starvation issues). This also would fail
to prevent bad replay scenarios (see below for what I think is
one which would cause problems).

Question: what happens in the case of a rogue BeginTxn/Write for
a key which is then abandoned for final commit epoch, followed by
a rogue EndTxn, if the rogues arrive after the txn has been
committed and cleaned up? The rogues in concert could resurrect
the abandoned intent and commit it, as these two rogue batches
would agree on epochs, and there was previously no protection at
MVCC for committing intents with timestamps which are newer than
the commit timestamp.

All intent resolutions should update the read timestamp cache to
the timestamp of the txn. This guarantees that before a txn is GCd,
the timestamp caches for any of its abandoned intents will disallow
replays from timestamps earlier or equal to any possible replay of
an earlier epoch's `EndTransaction` command.

The txn ID in the timestamp cache should be set to nil if a txn is
resolving its own intents or else to the pusher txn ID in the event
that a push is responsible for the intent resolve. This allows the
pusher to proceed without increasing its commit timestamp and
possibly restarting.

TODO(tschottdorf): I removed the code to bail early on retry txn
status in txn_coord_sender.go. I think what we want instead is a
way to query through the SQL driver whether the txn is still
expected to be valid. We could just duplicate the logic that's in
`EndTransaction`.

spencerkimball committed:
Fix slow sql TestParallel test via properly setting Txn.Writing

We previously were not propagating `Transaction.Writing` back to
the client, even after having written the transaction record on
`BeginTransaction`. The sql `TestParallel` test would restart
using a different first key, which would generate a new txn
record, leaving the old intents in place and no way to push the
old txn if it had a high enough priority short of waiting for it
to expire.

This commit always sets `Writing=true` on `BeginTransaction()`,
but takes care to restore the original value of `Writing` in the
event that a batch request has an error.

References #2920
Fixes #4442
Fixes #5376

spencerkimball committed:
Respond to comments, change replay protection mechanism

The previous replay protection mechanism was relying on the
read timestamp cache getting pushed by intent resolves.
Unfortunately, that was not protection against the following
scenario:
```
1) txn epoch 0
   1a) begin txn
   1b) put(A)
   1c) end txn: RETRY
2) txn epoch 1
   2a) begin txn
   2b) put(B)
   2c) end txn: SUCCEED
3) replay
   3a) replay request 1a
   3b) replay request 1b (gets higher timestamp)
   3c) read of A pushes txn epoch 0 created by 3a forward past 3b's timestamp
   3d) replay request 1c ends txn and successfully pushes A's intent
```

The new mechanism relies on putting EndTransaction calls into
the write timestamp cache so they trigger a WriteTooOld error
on replays. This guarantees that the EndTransaction request
must retry, which it won't on a replay.

Added missing unittests.

Commit:
Fix help us flow

- Turned the "help us" banner on by default.
- Fixed an issue where the help us page was resetting as people entered in form data.
- Added the real privacy policy.

spencerkimball committed:
Replay protection on Raft replays and deadlock issues

Raft replay protection via WriteTooOldError was lost with the
move to handling WriteTooOldErrors at the `executeBatch` level, but
this restores it in part by propagating the WriteTooOldError in the
event it occurs with an identical timestamp and there's not prior
overlap in the same batch (this allows a preceding DeleteRange or
preceding Increments, for example). This fixes a lot of unittests
which rely on Increment having no replays when sent non-transactionally.
There's no guarantee we're supposed to provide here, but it's probably
better to at least try, especially for the wealth of unittests which
rely on it. This change will always work for simple batches but might
allow replays in the event that a batch has overlapping keys.

Fixed some truly hairy deadlock issues with transactions which restart
with different keys for their first puts. This is what the sql
`partestdata/subquery_retry/txn` test does and running it in stress
mode really highlighted numerous problem cases. In particular:

- Certain conditions can allow two txns with interlocking put requests
  to find themselves with identical priorities (due to pushes) and
  timestamps (also due to pushes) such that neither can push the other
  because each has `timestamp.Next()` on the restart meaning each is
  unable to push the other.

  To solve, we now also sort on txn ID.

- Further conditions are possible which can push a txn's state either
  in priority or timestamp, making that txn able to push an intent it's
  waiting on if it knew it's own state. Unfortunately, if it's locked
  into a `WriteIntentError` backoff/retry loop, there was no way for it
  to become appraised of its new state.

  To solve, this change makes a txn query its state on backoff.

danhhz committed:
TxnCoordSender's heartbeat uses Context cancellation when available

Previously, the heartbeat was purely based on timeouts. Now, when given a
context.Context with a non-nil Done channel (which indicates that it is
cancellable), use that instead.

Plumb a cancellable Context from a pgwire connection down to this heartbeat.

Fixes #3031.

dt committed:
server: optionally send periodic usage reports

Second part of #5295, namely the per-node sending
of top-line node metrics (keys, bytes, ranges) for
deployments that opt-in to sharing via the "Help Us"
page in the UI.

Fixes #5295.

Commit:
first revisions

Commit:
revising Get In Touch and Contributing

Commit:
fixing indentation problems

Commit:
seeign if h1 header works

spencerkimball committed:
Further iteration on the replay protection to avoid blocking intents

This problem would come up in the TestSingleKey acceptance test on
occasion. The issue was that we allowed `BeginTransaction` requests
to succeed on replays and would fail `EndTransaction` requests with
a retry error. That would still leave the txn record in a `PENDING`
state, which if there were any intents written with the replay, would
need to be successfully pushed. When the TestSingleKey acceptance
test failed, it was because the contending transaction didn't have
the priority to push the intent so would timeout.

Now we set the write timestamp cache on `EndTransaction` but check
it on `BeginTransaction`. This sends a retry error on begin, meaning
there will be no txn record on end, so both will fail and any intents
which are established will be trivially pushed by any priority.

Also, replaced bad usage of `PUSH_TOUCH` when calling `PushTxn()`
from the store's backoff/retry `Send` loop on `WriteIntentErrors`
with `PUSH_UPDATE` which has the right semantics and restored the
old semantics for `PUSH_TOUCH`.

Commit:
making h1 a link

danhhz committed:
Change the encoding of durations to avoid fidelity loss

Durations were previously encoded the same way golang does, as an int64 count of
nanoseconds. But it's not possible to use this to represent something like "1
year 3 months" without either context or losing some information.

Postgres stores durations as a count of months, days, and fractions of seconds.
This works because there are always 12 months in a year, but the number of days
in a month varies. The number of seconds in a day also varies because of
daylight savings time. When sorting these durations, Postgres assumes that 1
month = 30 days and 1 day = 24 * 60 * 60 seconds. This means that 29d sorts
before 1m which sorts before 31d.

With inspiration from postgres, this change updates the internal representation
of durations to be a count of months, days, and nanoseconds. The encoding used
in sql tables matches the postgres sorting.

bdarnell committed:
Merge pull request #5487 from bdarnell/buildutil

testutils: move build.go into a subpackage

Commit:
client drivers and talks

nvb commented on roachpb/data.go:
```diff
@@ -436,6 +437,16 @@ func (v *Value) SetTime(t time.Time) {
+	// TODO(dan): BEFORE MERGE Figure out the actual number for encodingSizeOverestimate.
+	const encodingSizeOverestimate = 33
+	v.RawBytes = make([]byte, headerSize, headerSize+encodingSizeOverestimate)
+	v.RawBytes = encoding.EncodeDurationAscending(v.RawBytes, t)
```
The encoding for `roachpb.Value` does not need to sort correctly, so you might be able to speed this up or make the encoding smaller.

nvb commented on util/duration/duration.go:
```diff
@@ -0,0 +1,76 @@
+type Duration struct {
+	Months int
+	Days   int
+	Nanos  int64
```
Why `int` vs `int64`?

Commit:
minor changes to design overview

Commit:
minor improvements

Commit:
minor fixes

Commit:
header size change

tbg committed:
mvcc: always fail when SplitKey = StartKey

Previous code would return the descriptor's StartKey as valid SplitKey if there
was a value at StartKey. It isn't supposed to do that since the range is
already split at that point. See #5501.

cuongdo committed:
Make admin UI perform inconsistent reads

Also made the following changes:
- Moved events API call to the events page. Previously, we were always
  hitting the events API, which issues SQL queries that just hang when
  we lose our quorum.
- Marked the request to the uidata API as a background request, to
  prevent the Mithril event loop from blocking on it.

Resolves #4617.

cuongdo committed:
Add inconsistent read support to KV client

Added GetInconsistent, ScanInconsistent, GetProtoInconsistent, and a new
ReadConsistency field to client.Batch.

Commit:
optionally serve postgres on unix socket.

Fixes #5469

This is mainly needed for things like the testserver-from-real-binaries
which needs dynamic ports, but those are hard, it's easier to create a
temporary file.

Note: unix sockets in postgres are weird. If you pass
`--host=/var/lib/postgresql`, the actual socket file is
`/var/lib/postgresql/.s.PGSQL.5432`.

So when using existing clients (including the go libpq), you need to run it as follows:
```
$ ./cockroach start --insecure --socket=/tmp/.s.PGSQL.12345
$ ./cockroach sql --insecure --url=postgresql://:12345/?host=%2Ftmp%2F
```

Since this is so weird, I'm hidding the flag for now.

tbg committed:
storage: Factor out RangeDescriptor iteration

tbg committed:
Merge pull request #5510 from tschottdorf/splitkey

mvcc: always fail when SplitKey = StartKey

Commit:
Fix Time Series Merge Inconsistency

There was a bug in the Time Series system where the "Max" value for time series
could be different if an accumulated value was "merged" multiple times.

The specific bug was that we were using `std::numeric_limits<double>::min()` as if
it represented the lowest finite value (a very large negative number); however,
for floating point values, `numeric_limits::min` represents the smallest
  _positive_ value. Because of this, when merging an existing sample with an
  explicit max, _if_ the existing max was _negative_ then the new explicit max
  was being set to `std::numeric_limits<double>::min()`, which was unintended.

The simplest fix was to use `std::numeric_limits::lowest()`, which does provide
the lowest finite value as expected. However, we have opted instead to optimize
the situation where the destination message is empty in a call to
`AccumulateTimeSeriesSamples(dst, src)`. This obviates the need for a call to
std::numeric_limits, as there is not longer a situation where an explicit
maximum (or minimum) value is taken from an empty message.

This was discovered by the replica consistency checker - the "max" value for
some samples was different when the value was read from different replicas,
because the RocksDB merge operator had been executed more times on one replica
than the other.

Fixes #5414

tbg committed:
storage: Pass logger into MVCCFindSplitKey

tbg committed:
cli: Add cli/debug split-key <rangeID>

For #5501. The result of running it against #5501 is a little boring:

```
./cockroach debug split-key ~/5501/data.ec2-xx-xxx-xxx-121.compute-1.amazonaws.com/ 54
[...]
better split: diff 14161 at /System/StatusNode/4/1458696823.168982825,0
better split: diff 6542 at /System/StatusNode/4/1458696813.172379557,0
better split: diff 1077 at /System/StatusNode/4/1458696803.185926328,0
No SplitKey found: storage/engine/mvcc.go:1821: the range cannot be split; considered range "/System/StatusNode/4"-"/System/\"tsd\\x12cr.node.exec.latency-10m-p90\\x00\\x01\\x89\\xf8\\x06.\\xc13\"" has no valid splits
```

So essentially we just happen to have a Range whose StartKey has so many
versions that it's chosen as the SplitKey, which isn't legal.

tbg committed:
Merge pull request #5516 from tschottdorf/splitkey

cli: Add cli/debug split-key <rangeID>

knz committed:
Move the time monotonicity checker to the HLC.

mberhault committed:
Merge pull request #5513 from cockroachdb/marc/add_unix_socket

optionally serve postgres on unix socket.

mrtracy committed:
Merge pull request #5515 from mrtracy/mtracy/time_series_inconsistency

Fix Time Series Merge Inconsistency

vivekmenezes committed:
rocksdb: remove dead code WillOverflow()

spencerkimball committed:
WriteTooOldError propagation for raft replay protection on transactions

Extended the raft replay protection to transactions. Including
transactions means we can avoid leaving messy intents in case we
replay a BeginTxn / Put / EndTxn batch. The EndTxn would never
commit due to the WriteTooOld bool getting set, but this way we
avoid leaving an intent which may require a deadline expiration
to clear. This was one of the ways in which the TestSingleKey
acceptance test could fail.

Because it's possible for concurrent transactions which have backed
up on a single key to synchronize on their restart timestamps, we
still need to make sure that a propagated WriteTooOld error causes
restarts as opposed to just failing. We should avoid synchronizing
the timestamps of contending transactions, but that's another PR.

vivekmenezes committed:
storage: set version on all consistency checksum requests

Also set version to the non default value of 1.

nvb committed:
storage: Add a TimestampCache minimum size for eviction threshold

As discussed, a minimum ts cache size before beginning evictions will
avoid transactions driven by an interactive shell from being aborted
after only 10 seconds. The threshold size of 512 is arbitrary, but it
sounds like a good number for now.

RaduBerinde committed:
Temporarily increase test timeout to avoid test failures

Issue #5376 is causing a lot of test failures due to timeouts in the sql tests.
Temporarily increasing the timeouts until the issue is resolved.

RaduBerinde committed:
Merge pull request #5528 from RaduBerinde/increase-timeout

Temporarily increase test timeout to avoid test failures

RaduBerinde committed:
storage: document mu.desc as pointer-to-immutable

Documents the issue pointed out in #5509.

cuongdo committed:
Merge pull request #5519 from cuongdo/fix_admin_loss_of_quorum

Improve performance of admin UI under loss of quorum

Commit:
revisions for ben and spencer

Commit:
Merge pull request #5511 from jseldess/jseldess/readme-updates

Readme updates - fixes #5182

RaduBerinde committed:
Merge pull request #5529 from RaduBerinde/document-mu-desc

storage: document mu.desc as pointer-to-immutable

Commit:
Update README.md

minor correction

cuongdo committed:
ui: allow SetUIData to set/update multiple keys

This changes SetUIData so that instead of taking a single key & value
to store in system.uidata, it now takes a map of keys to values. This
allows the admin UI to set multiple pieces of uidata with a single HTTP
round-trip.

One motivation for this change is to allow the admin UI to store several
pieces of data individually, so that the server can access specific
fields without having to parse JSON.

maxlang committed:
Merge pull request #5485 from maxlang/maxlang/ui-fix-helpus

"Help Us" request on by default, and fixes

andreimatei committed:
sql: change the syntax for txn retries to SAVEPOINT

This changes the statements used to the standard sql sub-transaction
syntax:
SAVEPOINT cockroach_restart
RELEASE SAVEPOINT cockroach_restart
ROLLBACK TO SAVEPOINT cockroach_restart

The plan was always to probably support this syntax too since Ben argued
it helps with ORMs caching mechanism. I was forced to do this right away
because the C++ postgres driver (libpqxx) doesn't allow you to use a
transaction in any way after an error is returned, so there's no
oportunity to restart. But it does allow you to use subtransactions,
which is how they recommend doing error handling.

spencerkimball committed:
Merge pull request #5432 from cockroachdb/spencerkimball/always-write-intents

Always write intents on WriteTooOldError to prevent starvation

RaduBerinde committed:
Revert "Temporarily increase test timeout to avoid test failures"

RaduBerinde committed:
Merge pull request #5534 from cockroachdb/revert-5528-increase-timeout

Revert "Temporarily increase test timeout to avoid test failures"

dt committed:
Merge pull request #5496 from dt/reg

server: call home to check for new versions and optionally report metrics

tbg committed:
storage: Avoid ad-hoc mutation

Equivalent code, but Header might end up on the heap and it is more transparent
this way either way.

tbg committed:
roachpb: rename PUSH_{UPDATE->QUERY}, fix comment

We can't (and didn't) return the PusheeTxn verbatim because the parameter is
a TxnMeta and the return is a full-fledged Txn.

tbg committed:
storage: Touch up TestRaftReplayProtection

Verify the result of self-overlapping increments in a Batch.

knz committed:
Merge pull request #5522 from knz/hlc-monotonic-test

Move the time monotonicity checker to the HLC.

knz committed:
util: non-functional cosmetic changes to please the test machinery.

- Fix a spelling error in a comment.
- Ensure that time_fake_test imports log so its stays
  sensitive to the -verbosity flag.

tbg committed:
storage: unflake TestRaftReplayProtectionInTxn

An actual replay preceding the fake replay could leave a sequence cache
entry which altered the returned error.

Fixes #5536.

nvb committed:
storage: Remove MVCC value allocation on Put

nvb committed:
Merge pull request #5526 from nvanbenschoten/nvanbenschoten/tsthresh

storage: Add a TimestampCache minimum size for eviction threshold

tamird committed:
Merge pull request #5541 from knz/fix-spelling

Fix some test breakage.

Commit:
Fix "use strict" error in admin UI

A javascript statement was running afoul of "use strict" rules on firefox and
safari; apparently, this particular rule was not being enforced in chrome.

Rule in question: "Strict mode does not allow function declarations in a
lexically nested statement."

Fixes #5464

mrtracy committed:
Merge pull request #5548 from mrtracy/mtracy/fix_safari_firefox

Fix "use strict" error in admin UI

nvb committed:
kv: Properly thread context with timeout through kv sendOne

The context was available, but we weren't using it before.

tbg committed:
storage: deflake TestScannerEmptyRangeSet

Time is never your friend.

Fixes #5536.

tbg committed:
Merge pull request #5549 from tschottdorf/fix_flake

storage: deflake TestScannerEmptyRangeSet

tbg committed:
Merge pull request #5537 from tschottdorf/brushup

storage: Minor fallout to #5432

nvb committed:
Merge pull request #5553 from nvanbenschoten/nvanbenschoten/nopointer

storage: Remove MVCC value allocation on Put

andreimatei committed:
Merge pull request #5535 from andreimatei/retry-snapshot

sql: change the syntax for txn retries to SAVEPOINT

vivekmenezes committed:
Merge pull request #5523 from vivekmenezes/vivek/lowest

rocksdb: remove dead code WillOverflow()

cuongdo committed:
Merge pull request #5532 from cuongdo/set_uidata_map

ui: allow SetUIData to set/update multiple keys

danhhz committed:
Merge pull request #5447 from paperstreet/ctx

TxnCoordSender's heartbeat uses Context cancellation when available

nvb committed:
Merge pull request #5552 from nvanbenschoten/nvanbenschoten/sendCtx

kv: Properly thread context with timeout through kv sendOne

vivekmenezes committed:
Merge pull request #5524 from vivekmenezes/vivek/tsd

storage: set version on all consistency checksum requests

vivekmenezes committed:
Revert "TxnCoordSender's heartbeat uses Context cancellation when available"

vivekmenezes committed:
Merge pull request #5568 from cockroachdb/revert-5447-ctx

Revert "TxnCoordSender's heartbeat uses Context cancellation when available"

dt committed:
server: bail out of update check on http error

Fixes #5570.

dt committed:
server: fix out-of-date comments in updates.go

dt committed:
Merge pull request #5571 from dt/npe

server: bail out of update check on http error

kkaneda committed:
storage: Remove a TODO comment on nil response on error

add{ReadOnly,Write}Cmd returns a non-nil batch response on error. We
shouldn't just look at a response when an error is non-nil.

dt committed:
Merge pull request #5572 from dt/commentsalwayslie

server: fix out-of-date comments in updates.go

kkaneda committed:
Merge pull request #5573 from cockroachdb/kaneda/nil_response

storage: Remove a TODO comment on nil response on error
